Refactor create security group logic

varikin opened the pull request:
Moving that logic the provider to the operation.

Commit:
Refactored upsert openstack security group operation.

Moved business logic out of client provider and into the operation. The
client provider now just is a proxy for individual calls to the
Openstack API.

drmaas commented on clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/securitygroup/UpsertOpenstackSecurityGroupAtomicOperation.groovy:
```diff
@@ -47,12 +57,38 @@ class UpsertOpenstackSecurityGroupAtomicOperation implements AtomicOperation<Voi
+    if (StringUtils.isNotEmpty(description.id)) {
+      task.updateStatus BASE_PHASE, "Looking up existing security group with id ${description.id}"
+      securityGroup = provider.getSecurityGroup(description.id)
+      if (securityGroup == null) {
```
I think we have some inconsistencies in how to handle null results - in the load balancer registration/deregistration provider methods, a provider exception is thrown if null is returned from the client.

Then, in the operation classes, you could just to a try catch over the whole method, catch provider exception, do task logging and throw an operation exception with a detailed message (the provider message will be a bit generic).

I can see delegating the null handling to the Operation class though, too.

The thing that should be fixed here regardless is that the method should only throw OpenstackOperationException exceptions. Right now as is it could throw a provider exception if the provider method fails, and an oepration exception if null is found from the provider.

varikin replied:
I noticed that difference in your PR. I will adjust mine to match.

derekolk replied:
Do we want to throw an operation exception on a missing security group?  I would think we would want to throw a 404 back so the client can probably display resource not found.  Feels like we need to have consistency here.

varikin replied:
How do we throw a 404 from here?

derekolk commented on clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy:
```diff
@@ -78,50 +77,69 @@ abstract class OpenstackClientProvider {
-
+  SecGroupExtension.Rule createSecurityGroupRule(String securityGroupId, IPProtocol protocol, String cidr, int fromPort, int toPort) {
     handleRequest(AtomicOperations.UPSERT_SECURITY_GROUP) {
+      client.compute().securityGroups().createRule(Builders.secGroupRule()
```
You should update with region.

varikin replied:
I am going to do that in the next PR and do update all the operations we don't have it in yet.

varikin replied:
Oh region is already in the base description, it is less work than I thought, handling it now.

derekolk commented on clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy:
```diff
@@ -78,50 +77,69 @@ abstract class OpenstackClientProvider {
-  void upsertSecurityGroup(String securityGroupId, String securityGroupName, String description, List<OpenstackSecurityGroupDescription.Rule> rules) {
-
+  SecGroupExtension.Rule createSecurityGroupRule(String securityGroupId, IPProtocol protocol, String cidr, int fromPort, int toPort) {
     handleRequest(AtomicOperations.UPSERT_SECURITY_GROUP) {
```
Are you planning on updating with new handleRequest method without operation name?

varikin replied:
That was in your PR, is that merged yet?

cfieber committed:
Eliminates injection of AmazonCloudProvider for cache key handling

This introduces a constant on AmazonCloudProvider instead of injecting it as a component all over the place to reference its id
attribute.

drmaas commented:
lgtm

derekolk committed:
Delete load balancer operation. (#670)

Rebasing and conflict resolving

Removed unused import

cfieber committed:
Merge pull request #675 from cfieber/aws_infrastructure_keys_cleanup

Eliminates injection of AmazonCloudProvider for cache key handling

drmaas commented:
You will need to rename the bean name in UpsertOpenstackSecurityGroupAtomicOperationConverter to:

upsertOpenstackSecurityGroupDescription

to avoid bean name conflicts.

varikin commented:
@drmaas I removed the name completely. It is just `@Component` now.

emjburns committed:
openstack clone server groups (#679)

varikin committed:
Delete a security group. (#672)

This included refactoring the OpenstackSecurityGroupDescription to be
UpsertSecurityGroupDescription since delete only needs the id, not all
the options of upsert.
